File: binding/src/multimedia/libqabstractvideobuffer.cpp
#include <QAbstractVideoBuffer>
#define WORKAROUND_INNER_CLASS_DEFINITION_QAbstractVideoBuffer__MapData
#include <QVideoFrameFormat>
#include <qabstractvideobuffer.h>
#include "libqabstractvideobuffer.h"
#include "libqabstractvideobuffer.hxx"

QAbstractVideoBuffer* QAbstractVideoBuffer_new() {
	 return new VirtualQAbstractVideoBuffer();
}

QAbstractVideoBuffer__MapData* QAbstractVideoBuffer_Map(QAbstractVideoBuffer* self, int mode) {
	return new QAbstractVideoBuffer::MapData(self->map(static_cast<QVideoFrame::MapMode>(mode)));
}

void QAbstractVideoBuffer_Unmap(QAbstractVideoBuffer* self) {
	self->unmap();
}

QVideoFrameFormat* QAbstractVideoBuffer_Format(const QAbstractVideoBuffer* self) {
	return new QVideoFrameFormat(self->format());
}

// Base class handler implementation
QAbstractVideoBuffer__MapData* QAbstractVideoBuffer_QBaseMap(QAbstractVideoBuffer* self, int mode) {
	auto* vqabstractvideobuffer = dynamic_cast<VirtualQAbstractVideoBuffer*>(self);
	if (vqabstractvideobuffer && vqabstractvideobuffer->isVirtualQAbstractVideoBuffer) {
vqabstractvideobuffer->setQAbstractVideoBuffer_Map_IsBase(true);
	return new QAbstractVideoBuffer::MapData(vqabstractvideobuffer->map(static_cast<QVideoFrame::MapMode>(mode)));
}
}

// Auxiliary method to allow providing re-implementation
void QAbstractVideoBuffer_OnMap(QAbstractVideoBuffer* self, intptr_t slot) {
	auto* vqabstractvideobuffer = dynamic_cast<VirtualQAbstractVideoBuffer*>(self);
	if (vqabstractvideobuffer && vqabstractvideobuffer->isVirtualQAbstractVideoBuffer) {
vqabstractvideobuffer->setQAbstractVideoBuffer_Map_Callback(reinterpret_cast<VirtualQAbstractVideoBuffer::QAbstractVideoBuffer_Map_Callback>(slot));
}
}

// Base class handler implementation
void QAbstractVideoBuffer_QBaseUnmap(QAbstractVideoBuffer* self) {
	auto* vqabstractvideobuffer = dynamic_cast<VirtualQAbstractVideoBuffer*>(self);
	if (vqabstractvideobuffer && vqabstractvideobuffer->isVirtualQAbstractVideoBuffer) {
vqabstractvideobuffer->setQAbstractVideoBuffer_Unmap_IsBase(true);
	vqabstractvideobuffer->unmap();
}
}

// Auxiliary method to allow providing re-implementation
void QAbstractVideoBuffer_OnUnmap(QAbstractVideoBuffer* self, intptr_t slot) {
	auto* vqabstractvideobuffer = dynamic_cast<VirtualQAbstractVideoBuffer*>(self);
	if (vqabstractvideobuffer && vqabstractvideobuffer->isVirtualQAbstractVideoBuffer) {
vqabstractvideobuffer->setQAbstractVideoBuffer_Unmap_Callback(reinterpret_cast<VirtualQAbstractVideoBuffer::QAbstractVideoBuffer_Unmap_Callback>(slot));
}
}

// Base class handler implementation
QVideoFrameFormat* QAbstractVideoBuffer_QBaseFormat(const QAbstractVideoBuffer* self) {
	auto* vqabstractvideobuffer = dynamic_cast<const VirtualQAbstractVideoBuffer*>(self);
	if (vqabstractvideobuffer && vqabstractvideobuffer->isVirtualQAbstractVideoBuffer) {
vqabstractvideobuffer->setQAbstractVideoBuffer_Format_IsBase(true);
	return new QVideoFrameFormat(vqabstractvideobuffer->format());
}
}

// Auxiliary method to allow providing re-implementation
void QAbstractVideoBuffer_OnFormat(const QAbstractVideoBuffer* self, intptr_t slot) {
	auto* vqabstractvideobuffer = dynamic_cast<const VirtualQAbstractVideoBuffer*>(self);
	if (vqabstractvideobuffer && vqabstractvideobuffer->isVirtualQAbstractVideoBuffer) {
vqabstractvideobuffer->setQAbstractVideoBuffer_Format_Callback(reinterpret_cast<VirtualQAbstractVideoBuffer::QAbstractVideoBuffer_Format_Callback>(slot));
}
}

void QAbstractVideoBuffer_Delete(QAbstractVideoBuffer* self) {
    delete self;
}

QAbstractVideoBuffer__MapData* QAbstractVideoBuffer__MapData_new(const QAbstractVideoBuffer__MapData* other) {
	 return new QAbstractVideoBuffer::MapData(*other);
}

QAbstractVideoBuffer__MapData* QAbstractVideoBuffer__MapData_new2(QAbstractVideoBuffer__MapData* other) {
	 return new QAbstractVideoBuffer::MapData(*other);
}

void QAbstractVideoBuffer__MapData_CopyAssign(QAbstractVideoBuffer__MapData* self, QAbstractVideoBuffer__MapData* other) {
    *self = *other;
}

void QAbstractVideoBuffer__MapData_MoveAssign(QAbstractVideoBuffer__MapData* self, QAbstractVideoBuffer__MapData* other) {
    *self = std::move(*other);
}

int QAbstractVideoBuffer__MapData_PlaneCount(const QAbstractVideoBuffer__MapData* self) {
	return self->planeCount;
}

void QAbstractVideoBuffer__MapData_SetPlaneCount(QAbstractVideoBuffer__MapData* self, int planeCount) {
	self->planeCount;
}

void QAbstractVideoBuffer__MapData_Delete(QAbstractVideoBuffer__MapData* self) {
    delete self;
}

File: binding/src/multimedia/libqabstractvideobuffer.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qabstractvideobuffer.html
pub const qabstractvideobuffer = struct {

    /// New constructs a new QAbstractVideoBuffer object.
    pub fn New() QtC.QAbstractVideoBuffer {
        return qtc.QAbstractVideoBuffer_new();
    }


    pub fn Map(self: ?*anyopaque, mode: i32) QtC.QAbstractVideoBuffer__MapData {
        return qtc.QAbstractVideoBuffer_Map(@ptrCast(self), @intCast(mode));
    }

    /// Allows for overriding the related default method
    pub fn OnMap(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) QtC.QAbstractVideoBuffer__MapData) void {
        qtc.QAbstractVideoBuffer_OnMap(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMap(self: ?*anyopaque, mode: i32) QtC.QAbstractVideoBuffer__MapData {
        return qtc.QAbstractVideoBuffer_QBaseMap(@ptrCast(self), @intCast(mode));
    }

    pub fn Unmap(self: ?*anyopaque, ) void {
        qtc.QAbstractVideoBuffer_Unmap(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnUnmap(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QAbstractVideoBuffer_OnUnmap(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseUnmap(self: ?*anyopaque, ) void {
        qtc.QAbstractVideoBuffer_QBaseUnmap(@ptrCast(self));
    }

    pub fn Format(self: ?*anyopaque, ) QtC.QVideoFrameFormat {
        return qtc.QAbstractVideoBuffer_Format(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnFormat(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QVideoFrameFormat) void {
        qtc.QAbstractVideoBuffer_OnFormat(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseFormat(self: ?*anyopaque, ) QtC.QVideoFrameFormat {
        return qtc.QAbstractVideoBuffer_QBaseFormat(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QAbstractVideoBuffer_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qabstractvideobuffer-mapdata.html
pub const qabstractvideobuffer__mapdata = struct {

    /// New constructs a new QAbstractVideoBuffer::MapData object.
    pub fn New(other: ?*anyopaque) QtC.QAbstractVideoBuffer__MapData {
        return qtc.QAbstractVideoBuffer__MapData_new(@ptrCast(other));
    }


    /// New2 constructs a new QAbstractVideoBuffer::MapData object.
    pub fn New2(other: ?*anyopaque) QtC.QAbstractVideoBuffer__MapData {
        return qtc.QAbstractVideoBuffer__MapData_new2(@ptrCast(other));
    }

    /// CopyAssign shallow copies `other` into `self`.
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QAbstractVideoBuffer__MapData_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QAbstractVideoBuffer__MapData_MoveAssign(@ptrCast(self), @ptrCast(other));
    }


    pub fn PlaneCount(self: ?*anyopaque, ) i32 {
        return qtc.QAbstractVideoBuffer__MapData_PlaneCount(@ptrCast(self));
    }

    pub fn SetPlaneCount(self: ?*anyopaque, planeCount: i32) void {
        qtc.QAbstractVideoBuffer__MapData_SetPlaneCount(@ptrCast(self), planeCount);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QAbstractVideoBuffer__MapData_Delete(@ptrCast(self));
    }
};
